core/drivers/crypto/caam/caam_rng.c

```diff
@@ -0,0 +1,600 @@
+			atomic_store_u32(&rng->status, DATA_EMPTY);
+		} else {
+			/* There is enough data in the current context */
+			RNG_TRACE("Copy %zu data", remlen);
+			memcpy(rngbuf, &rng->data[rng->rdindex], remlen);
+			rng->rdindex += remlen;
+			remlen = 0;
+		}
+	} while (remlen);
+
+	return TEE_SUCCESS;
+}
+
+/*
+ * Initialize the RNG module to generate data
+ *
+ * @ctrl_addr   Controller base address
+ */
+static enum CAAM_Status caam_rng_init_data(void)
+{
+	enum CAAM_Status retstatus = CAAM_FAILURE;
+
+	struct rngdata *rng;
+	uint8_t idx;
+
+	for (idx = 0; (idx < RNG_DATABUF_NB); idx++) {
+		rng = &rng_privdata->databuf[idx];
+		retstatus = prepare_gen_desc(rng);
+
+		if (retstatus != CAAM_NO_ERROR)
+			break;
+	}
+
+	return retstatus;
+}
+#endif /* CFG_CRYPTO_RNG_HW */
+
+/*
+ * Prepares the instantiation descriptor
+ *
+ * @nb_sh      Number of the State Handle
+ * @sh_status  State Handles status
+ * @desc       Reference to the descriptor
+ * @desc       [out] Descriptor filled
+ */
+static void prepare_inst_desc(uint32_t nb_sh, uint32_t sh_status,
+			      uint32_t *desc)
+{
+	bool key_loaded = false;
+	uint8_t sh_idx = 0;
+	uint8_t nb_max_sh = nb_sh;
+
+	/* Read the SH and secure key status */
+	key_loaded = caam_hal_rng_key_loaded(rng_privdata->baseaddr);
+	RNG_TRACE("RNG SH Status 0x%08" PRIx32 " - Key Status %d", sh_status,
+		  key_loaded);
+
+	while (sh_status & (1 << sh_idx))
+		sh_idx++;
+
+	RNG_TRACE("Instantiation start at SH%d (%d)", sh_idx, nb_max_sh);
+
+	/* Don't set the descriptor header now */
+	caam_desc_init(desc);
+	caam_desc_add_word(desc, DESC_HEADER(0));
+	/* First State Handle to instantiate */
+	caam_desc_add_word(desc, RNG_SH_INST(sh_idx));
+
+	/* Next State Handle */
+	sh_idx++;
+
+	while (sh_idx < nb_max_sh) {
+		if (!(sh_status & (1 << sh_idx))) {
+			/*
+			 * If there is more SH to instantiate, add a wait loop
+			 * followed by a reset the done status to execute next
+			 * command
+			 */
+			caam_desc_add_word(desc,
+					   JUMP_C1_LOCAL(ALL_COND_TRUE,
+							 JMP_COND(NONE), 1));
+			caam_desc_add_word(desc,
+					   LD_NOCLASS_IMM(REG_CLEAR_WRITTEN,
+							  sizeof(uint32_t)));
+			caam_desc_add_word(desc, 0x1);
+			caam_desc_add_word(desc, RNG_SH_INST(sh_idx));
+		}
+		/* Next State Handle */
+		sh_idx++;
+	}
+
+	/* Load the Key if needed */
+	if (!key_loaded) {
+		/*
+		 * Add a wait loop followed by a reset the done status
+		 * to execute next command
+		 */
+		caam_desc_add_word(desc, JUMP_C1_LOCAL(ALL_COND_TRUE,
+						       JMP_COND(NONE), 1));
+		caam_desc_add_word(desc, LD_NOCLASS_IMM(REG_CLEAR_WRITTEN,
+							sizeof(uint32_t)));
+		caam_desc_add_word(desc, 0x1);
+		caam_desc_add_word(desc, RNG_GEN_SECKEYS);
+	}
+
+	RNG_DUMPDESC(desc);
+}
+
+/* Instantiates the RNG State Handles if not already done */
+enum CAAM_Status caam_rng_instantiation(void)
+{
+	enum CAAM_Status retstatus = CAAM_FAILURE;
+	struct caam_jobctx jobctx = { 0 };
+	uint32_t *desc = NULL;
+	uint32_t sh_status = 0;
```

Comment:
I don't prefer because this is linked to HW registers which are 32 bits width.

Reply:
I think that in the patch there's a tendency to use `uint32_t` as the default type here and there. Fixed width types should generally only be used when they are really needed or you may impose unnecessary restrictions on the code. It may also make it harder to maintain the code since it sometimes takes an effort to realize that a particular property of `uint32_t` is really depended on.

I'll stop nagging about this (in my mind) frivolous use of `uint32_t` in this driver from now on.